### ISSUE TITLE
Adjoint hessian method

### ISSUE THREAD

**Issue by AuroraFr** [patrick-kidger/diffrax#628]
Hello,
Thanks a lot for sharing this lib.
I wonder if you used adjoint method to calculate hessian as calculating gradient ? This adjoint hessian method could be more efficient than jax.hessian ?
Thanks for your reply in advance.
Best,

**Comment by johannahaffner** [comment]
Hi Aurora, 

the adjoints enable differentiation through an ODE solve, they do not perform it. `jax.hessian` is one of several ways to obtain a Hessian of an arbitrary function in JAX, which may or may not involve an ODE solve. It will perform forward-over-reverse mode automatic differentiation, which would limit you to using diffrax' `DirectAdjoint`. You can also do `jax.jacfwd(jax.jacfwd(..))`, or the equivalent in reverse-mode, to use `diffrax.RecursiveCheckpointAdjoint` (only reverse-mode) or `diffrax.ForwardMode` instead.

**Comment by AuroraFr** [comment]
Hi, 
Thanks for your quick reply.
I used jax.hessian with DirectAdjoint to get fisher information matrix and then the variance. But in my case, it's too long to calculate and i got sometimes negative variance . I wonder if i implement adjoint hessian. It could help ? I would like your comments on this idea. 
Thanks !

**Comment by AuroraFr** [comment]
Hi, 
Thanks a lot for your reply. 
I think the negative variances are due to numerical errata. In my case, i need to do sampling before calculating the hessian because i'm in a incomplete data case. I didn't get always negative variances for the same dataset among several runs. 
I will look at how the forwardmode works. 
I saw your example, i use saveat so for the moment i use DirectAdjoint.

**Comment by johannahaffner** [comment]
Hi Aurora, 

Do you sample the parameters from the log-likelihood? Then you could also compute the sample covariance directly.
If you're sampling extra data points then it makes sense that your Hessian would be impacted, since positive-definiteness of estimated covariance matrices is only reasonable at the maximum likelihood estimate.

**Comment by AuroraFr** [comment]
I sample the latent data from estimated latent distribution.
